packages/vite/src/node/server/environments/fetchableEnvironments.ts

```diff
@@ -0,0 +1,59 @@
+import type { ResolvedConfig } from '../../config'
+import type { DevEnvironmentContext } from '../environment'
+import { DevEnvironment } from '../environment'
+import type { Environment } from '../../environment'
+
+export interface FetchableDevEnvironmentContext extends DevEnvironmentContext {
+  handleRequest(request: Request): Promise<Response> | Response
+}
+
+export function createFetchableDevEnvironment(
+  name: string,
+  config: ResolvedConfig,
+  context: FetchableDevEnvironmentContext,
+): FetchableDevEnvironment {
+  if (typeof Request === 'undefined' || typeof Response === 'undefined') {
+    throw new TypeError(
+      'FetchableDevEnvironment requires a global `Request` and `Response` object.',
+    )
+  }
+
+  if (!context.handleRequest) {
+    throw new TypeError(
+      'FetchableDevEnvironment requires a `handleRequest` method during initialisation.',
+    )
+  }
+
+  return new FetchableDevEnvironment(name, config, context)
+}
+
+export function isFetchableDevEnvironment(
+  environment: Environment,
+): environment is FetchableDevEnvironment {
+  return environment instanceof FetchableDevEnvironment
+}
+
+class FetchableDevEnvironment extends DevEnvironment {
+  private _handleRequest: (request: Request) => Promise<Response> | Response
+
+  constructor(
+    name: string,
+    config: ResolvedConfig,
+    context: FetchableDevEnvironmentContext,
+  ) {
+    super(name, config, context as DevEnvironmentContext)
```

Comment:
```suggestion
    super(name, config, context)
```
I think this type assertion is not needed as `FetchableDevEnvironmentContext` extends `DevEnvironmentContext`.